theano/gpuarray/opt.py

```diff
@@ -210,6 +211,77 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            if type(i) is not theano.tensor.TensorVariable:
+                continue
+            mapping[i] = GpuFromHost(None)(i)
+
+        for n in fgraph.toposort():
+            for o in n.outputs:
+                if type(o) is not theano.tensor.TensorVariable:
+                    continue
+                mapping[o] = GpuFromHost(None)(o)
+
+        for node in fgraph.toposort():
+
+            # The Extra condition
+            if any([isinstance(i, GpuArrayVariable) or
+               isinstance(i, GpuArraySharedVariable)
+               for i in node.inputs + node.outputs]):
+
+                move_to_GPU = False
+
+            # Oplifter's condition
+            # Will return a list of OP
+            # If None, means can't be moved.
+
+            new_ops = []
+            for lopt in (gpu_optimizer.query().local_optimizers_all +
+                         gpu_optimizer.query().local_optimizers_map.get(type(node.op), []) +
+                         gpu_optimizer.query().local_optimizers_map.get(node.op, [])):
+
+                new_ops.append(lopt.transform(node) or lopt(node))
+
+            if all(isinstance(x, Op) for x in new_ops):
+                move_to_GPU = False
+
+            if not new_ops:
+                move_to_GPU = False
+
+            for i in node.inputs:
+                if type(i) is not theano.tensor.TensorVariable:
+                    continue
+                newnode = node.clone_with_new_inputs([mapping.get(i) for i in node.inputs])
+
+            if move_to_GPU:
```

Comment:
Here you should create a new node like:

newnode=new_ops(*[mapping.get(i) for i in node.inputs])